core/src/env/srml/srml_only/impls.rs

```diff
@@ -152,20 +152,25 @@ where
         gas: <Self as EnvTypes>::Gas,
         value : <Self as EnvTypes>::Balance,
         input_data : &[u8]
-    ) -> u32 {
+    ) -> Option< Vec<u8> > {
```

Comment:
Returning a `Vec<u8>` directly is exactly what I warned about since then all success path do a costly read to the scratch buffer and even if the caller doesn't even use it. Also the compiler is probably not powerful enough to actually optimize this call away since it is over the FFI.
What we actually want is a type abstraction that does the trick of reading the scratch buffer.

E.g.:

```rust
struct ReturnData<T> {
     marker: PhantomData<fn () -> T>,
}

impl<T> ReturnData<T>
where
    T: parity_codec::Decode,
{
    fn get() -> T {
        T::decode(&mut &read_scratch_buffer()[..])
    }
}
```

This way we have an API that can only be used to load from the scratch buffer if `Some(_)` is returned. Also it leaves control over loading the actual return data to the caller and also directly decodes it to the underlying return type of the called function. Later we will have to wrap this function in order to build the more high-level alternative that is also type safe with respect to the called function's return type.

Reply:
Should I decode the data in this low level api?
To do this, it requires to have another generic type in Env traits.
Also, i'm not sure where to implement this wrapping structure.